src/server/scripts/EasternKingdoms/ZulGurub/boss_thekal.cpp

```diff
@@ -93,48 +108,139 @@ class boss_thekal : public CreatureScript
-                events.ScheduleEvent(EVENT_SILENCE, 9s, 0, PHASE_ONE);          // Phase 1
-                events.ScheduleEvent(EVENT_CHECK_TIMER, 10s, 0, PHASE_ONE);     // Phase 1
-                events.ScheduleEvent(EVENT_RESURRECT_TIMER, 10s, 0, PHASE_ONE); // Phase 1
-                Talk(SAY_AGGRO);
+                events.SetPhase(PHASE_ONE);
+                events.ScheduleEvent(EVENT_MORTALCLEAVE, 4s, 0, PHASE_ONE);
+                events.ScheduleEvent(EVENT_SILENCE, 9s, 0, PHASE_ONE);
             }
 
-            void JustReachedHome() override
+            void SetData(uint32 type, uint32 data) override
             {
-                instance->SetBossState(DATA_THEKAL, NOT_STARTED);
+                if (type == DATA_FAKE_DEATH)
+                {
+                    // A mob died
+                    switch (data)
+                    {
+                        case NPC_HIGH_PRIEST_THEKAL:
+                            _isThekalDead = true;
+                            break;
+                        case NPC_ZEALOT_LORKHAN:
+                            _isLorkhanDead = true;
+                            break;
+                        case NPC_ZEALOT_ZATH:
+                            _isZathDead = true;
+                            break;
+                    }
+
+                    if (_isThekalDead && _isLorkhanDead && _isZathDead)
+                    {
+                        _isEvadeDisabled = true;
+                        me->SetControlled(true, UNIT_STATE_ROOT);
+                        events.Reset();
+                        _isResurrectTimerActive = false;
+                        DoCastSelf(SPELL_RESURRECT);
+                        me->RemoveAurasDueToSpell(SPELL_PERMANENT_FEIGN_DEATH);
+                        me->SetFullHealth();
+                        events.ScheduleEvent(EVENT_CHANGE_PHASE_1, 1s);
+                    }
+                    else
+                    {
+                        // Start resurrection timer if not already started, otherwise ignore
+                        if (!_isResurrectTimerActive)
+                        {
+                            events.ScheduleEvent(EVENT_RESURRECT_TIMER, 10s);
+                            _isResurrectTimerActive = true;
+                        }
+                    }
+                }
+                else if (type == DATA_RESURRECTED)
+                {
+                    Creature* creature = nullptr;
+                    if (data == NPC_HIGH_PRIEST_THEKAL)
+                        creature = me;
+                    else if (data == NPC_ZEALOT_LORKHAN)
+                        creature = instance->GetCreature(DATA_LORKHAN);
+                    else if (data == NPC_ZEALOT_ZATH)
+                        creature = instance->GetCreature(DATA_ZATH);
+
+                    // Resurrect
+                    creature->RemoveAurasDueToSpell(SPELL_PERMANENT_FEIGN_DEATH);
```

Comment:
missing check if creature pointer is valid?
Can happens if instance cant find the creature.

Reply:
Added asserts, it's guaranteed that those creatures will exist at this point in the script unless something very bad is going on.

Reply:
Do they exist at this point even if one calls ".npc delete" on them halfway through the fight ?

Reply:
It will crash if a GM deletes the creature after it begins feigning death and before it is resurrected.

Didn't think about harmful GMs, going to remove asserts and silently fail in that case. Thanks :P